src/main/java/com/commercetools/sync/states/StateSync.java

```diff
@@ -0,0 +1,445 @@
+            .handle(ImmutablePair::new)
+            .thenCompose(updateResponse -> {
+                final Throwable sphereException = updateResponse.getValue();
+
+                if (sphereException != null) {
+                    return executeSupplierIfConcurrentModificationException(
+                        sphereException,
+                        () -> fetchAndUpdate(oldState, newState),
+                        () -> {
+                            final String errorMessage =
+                                format(CTP_STATE_UPDATE_FAILED, newState.getKey(),
+                                    sphereException.getMessage());
+                            handleError(errorMessage, sphereException, 1);
+                            return completedFuture(null);
+                        });
+                } else {
+                    statistics.incrementUpdated();
+                    return completedFuture(null);
+                }
+            });
+    }
+
+    @Nonnull
+    private CompletionStage<Void> fetchAndUpdate(
+        @Nonnull final State oldState,
+        @Nonnull final StateDraft newState) {
+        String key = oldState.getKey();
+        return stateService
+            .fetchState(key)
+            .handle(ImmutablePair::new)
+            .thenCompose(fetchResponse -> {
+                Optional<State> fetchedStateOptional = fetchResponse.getKey();
+                final Throwable exception = fetchResponse.getValue();
+
+                if (exception != null) {
+                    final String errorMessage = format(CTP_STATE_UPDATE_FAILED, key,
+                        "Failed to fetch from CTP while retrying after concurrency modification.");
+                    handleError(errorMessage, exception, 1);
+                    return completedFuture(null);
+                }
+
+                return fetchedStateOptional
+                    .map(fetchedState -> buildActionsAndUpdate(fetchedState, newState))
+                    .orElseGet(() -> {
+                        final String errorMessage = format(CTP_STATE_UPDATE_FAILED, key,
+                            "Not found when attempting to fetch while retrying "
+                                + "after concurrency modification.");
+                        handleError(errorMessage, null, 1);
+                        return completedFuture(null);
+                    });
+            });
+    }
+
+    @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION") // https://github.com/findbugsproject/findbugs/issues/79
+    @Nonnull
+    private CompletionStage<Void> resolveNowReadyReferences() {
+
+        // We delete anyways the keys from the statistics before we attempt resolution, because even if resolution fails
+        // the states that failed to be synced would be counted as failed.
+
+        final Set<String> referencingDraftKeys = readyToResolve
+            .stream()
+            .map(statistics::removeAndGetReferencingKeys)
```

Comment:
Does it give you the list of the state keys that have been created during current batch only or all of the processed state keys?

Reply:
It returns the keys of all processed keys